src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/finance/TransactionSynchronisationService.kt

```diff
@@ -23,17 +26,44 @@ class TransactionSynchronisationService(
   private val financeService: FinanceApiService,
   private val telemetryClient: TelemetryClient,
   private val queueService: SynchronisationQueueService,
+  private val transactionIdBufferRepository: TransactionIdBufferRepository,
+  @Value("\${finance.transactions.forwardingDelaySeconds}")
+  private val forwardingDelaySeconds: Int,
 ) {
   private companion object {
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  suspend fun transactionInserted(event: TransactionEvent, requestId: UUID) {
+  suspend fun transactionInsertCheck(event: TransactionEvent, requestId: UUID) {
+    if (event.isSourcedFromDPS()) {
+      telemetryClient.trackEvent("transactions-synchronisation-created-skipped", event.toTelemetryProperties())
+      return
+    }
+
+    if (transactionIdBufferRepository.existsById(event.transactionId)) {
+      log.debug("Skipping synchronisation of transaction ${event.transactionId} as id already recorded")
+      return
+    }
+
+    if (transactionIdBufferRepository.addId(event.transactionId) == 0) {
+      log.debug("Skipping synchronisation of transaction ${event.transactionId} as failed to record id")
+      return
```

Comment:
Shouldn't this be an exception?

Reply:
This is the race condition where 2 threads try to insert at the same time and this is the one that fails. I think other errors would result in an exception from the repository method but ill double check ..